contracts/dao/neutron-chain-manager/src/contract.rs

```diff
@@ -0,0 +1,305 @@
+        }
+        ExecuteMsg::RemoveStrategy { address } => execute_remove_strategy(deps, info, address),
+        ExecuteMsg::ExecuteMessages { messages } => execute_execute_messages(deps, info, messages),
+    }
+}
+
+pub fn execute_add_strategy(
+    deps: DepsMut,
+    info: MessageInfo,
+    address: Addr,
+    strategy: Strategy,
+) -> Result<Response<NeutronMsg>, ContractError> {
+    is_authorized(deps.as_ref(), info.sender.clone())?;
+
+    // We add the new strategy, and then we check that it did not replace
+    // the only existing ALLOW_ALL strategy.
+    STRATEGIES.save(deps.storage, address.clone(), &strategy)?;
+    if let Strategy::AllowOnly(_) = strategy {
+        if get_allow_all_count(deps.as_ref())? == 0 {
+            return Err(ContractError::InvalidDemotion {});
+        }
+    }
+
+    Ok(Response::new()
+        .add_attribute("action", "execute_add_strategy")
+        .add_attribute("address", address))
+}
+
+pub fn execute_remove_strategy(
+    deps: DepsMut,
+    info: MessageInfo,
+    address: Addr,
+) -> Result<Response<NeutronMsg>, ContractError> {
+    is_authorized(deps.as_ref(), info.sender.clone())?;
+
+    // First we remove the strategy, then we check that it was not the only
+    // ALLOW_ALL strategy we had.
+    STRATEGIES.remove(deps.storage, address.clone());
+    if get_allow_all_count(deps.as_ref())? == 0 {
+        return Err(ContractError::InvalidDemotion {});
+    }
+
+    Ok(Response::new()
+        .add_attribute("action", "execute_remove_strategy")
+        .add_attribute("address", address))
+}
+
+pub fn execute_execute_messages(
+    deps: DepsMut,
+    info: MessageInfo,
+    messages: Vec<CosmosMsg<NeutronMsg>>,
+) -> Result<Response<NeutronMsg>, ContractError> {
+    // If the sender doesn't have a strategy associated with them, abort immediately.
+    if !STRATEGIES.has(deps.storage, info.sender.clone()) {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    let response = Response::new()
+        .add_attribute("action", "execute_execute_messages")
+        .add_attribute("address", info.sender.clone());
+
+    let strategy = STRATEGIES.load(deps.storage, info.sender)?;
+    match strategy {
+        Strategy::AllowAll => Ok(response
+            .add_attribute("strategy", "allow_all")
+            .add_messages(messages)),
+        Strategy::AllowOnly(_) => {
+            check_allow_only_permissions(deps.as_ref(), strategy.clone(), messages.clone())?;
+            Ok(response
+                .add_attribute("strategy", "allow_only")
+                .add_messages(messages))
+        }
+    }
+}
+
+fn is_authorized(deps: Deps, address: Addr) -> Result<(), ContractError> {
+    match STRATEGIES.load(deps.storage, address) {
+        Ok(strategy) => {
+            if let Strategy::AllowOnly(_) = strategy {
+                return Err(ContractError::Unauthorized {});
+            }
+
+            Ok(())
+        }
+        Err(_) => Err(ContractError::Unauthorized {}),
```

Comment:
it's a bit easier to read, from my point of view
```suggestion
        Ok(Strategy::AllowAll) => Ok(()),
        _ => Err(ContractError::Unauthorized {})
```

Reply:
didn't even know you could do this. thanks